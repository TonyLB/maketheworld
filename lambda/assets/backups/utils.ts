export const getCurrentDateString = (): string => {
    return new Date().toISOString().split('T')[0]
}
