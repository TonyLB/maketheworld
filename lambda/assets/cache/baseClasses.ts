import { ComponentRenderItem } from "@tonylb/mtw-wml/dist/normalize"

export type EphemeraCondition = {
    dependencies: string[];
    if: string;
}

export type EphemeraFeatureAppearance = {
    conditions: EphemeraCondition[];
    name: string;
    render: ComponentRenderItem[];
}

export type EphemeraFeature = {
    EphemeraId: string;
    key: string;
    tag: 'Feature';
    appearances: EphemeraFeatureAppearance[];
}

export type EphemeraExit = {
    name: string;
    to: string;
}

export type EphemeraRoomAppearance = {
    conditions: EphemeraCondition[];
    name: string;
    render: ComponentRenderItem[];
    exits: EphemeraExit[];
}

export type EphemeraRoom = {
    EphemeraId: string;
    key: string;
    tag: 'Room';
    appearances: EphemeraRoomAppearance[];
}

export type EphemeraMapRoom = {
    EphemeraId: string;
    x: number;
    y: number;
}

export type EphemeraMapAppearance = {
    conditions: EphemeraCondition[];
    fileURL: string;
    name: string;
    rooms: Record<string, EphemeraMapRoom>;
}

export type EphemeraMap = {
    EphemeraId: string;
    key: string;
    tag: 'Map';
    appearance: EphemeraMapAppearance;
}

//
// TODO (ISS1385): Add EphemeraAction to base classes
//

export type EphemeraItem = EphemeraFeature | EphemeraRoom | EphemeraMap
