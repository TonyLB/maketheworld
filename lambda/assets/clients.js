import { S3Client, GetObjectCommand as GetObjectCommandOriginal } from "@aws-sdk/client-s3"
import { EventBridgeClient } from "@aws-sdk/client-eventbridge"

export const ebClient = new EventBridgeClient({ region: process.env.AWS_REGION })

// export const s3Client = new S3Client(params)

export class GetObjectCommand extends GetObjectCommandOriginal {}
