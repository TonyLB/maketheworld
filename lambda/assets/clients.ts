import { GetObjectCommand as GetObjectCommandOriginal } from "@aws-sdk/client-s3"
import { EventBridgeClient, PutEventsCommand } from "@aws-sdk/client-eventbridge"
import { apiClient as apiClientImport } from "@tonylb/mtw-utilities/dist/apiManagement/apiManagementClient"
import { SFNClient } from "@aws-sdk/client-sfn"

export const ebClient = new EventBridgeClient({ region: process.env.AWS_REGION })
export const sfnClient = new SFNClient({ region: process.env.AWS_REGION })

// export const s3Client = new S3Client(params)

export class GetObjectCommand extends GetObjectCommandOriginal {}

export const apiClient = {
    send: (message: any) => {
        return apiClientImport.send(message)
            .catch(async (err: any) => {
                if (err.name === 'GoneException' || err.name === 'BadRequestException') {
                    await ebClient.send(new PutEventsCommand({
                        Entries: [{
                            EventBusName: process.env.EVENT_BUS_NAME,
                            Source: 'mtw.diagnostics',
                            DetailType: 'Force Disconnect',
                            Detail: JSON.stringify({
                                connectionId: message.ConnectionId
                            })
                        }]
                    }))
                }
                else {
                    console.log(`Error: ${err.name}`)
                    throw err
                }
            })
    }
}
