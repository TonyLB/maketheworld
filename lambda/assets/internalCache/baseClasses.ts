export class CacheBase {
    clear() {}
    async flush() {}
}

type Constructor<T = {}> = new (...args: any[]) => T;

export type CacheConstructor = Constructor<{
    clear(): void;
    flush(): Promise<void>;
}>
