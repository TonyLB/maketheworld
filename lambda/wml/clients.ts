import { EventBridgeClient } from "@aws-sdk/client-eventbridge"
import { S3Client } from "@aws-sdk/client-s3"
import { SFNClient } from "@aws-sdk/client-sfn"
import { SNSClient } from "@aws-sdk/client-sns"
import AWSXRay from 'aws-xray-sdk'

const params = { region: process.env.AWS_REGION }
export const ebClient = AWSXRay.captureAWSv3Client(new EventBridgeClient(params))
export const sfnClient = AWSXRay.captureAWSv3Client(new SFNClient(params))
export const snsClient = AWSXRay.captureAWSv3Client(new SNSClient(params))
export const s3Client = AWSXRay.captureAWSv3Client(new S3Client(params))
