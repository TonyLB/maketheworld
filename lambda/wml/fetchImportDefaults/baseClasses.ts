import { EphemeraAssetId } from "@tonylb/mtw-interfaces/ts/baseClasses"
import { Graph } from "@tonylb/mtw-utilities/dist/graphStorage/utils/graph"
import { AssetWorkspaceAddress } from "@tonylb/mtw-asset-workspace"
import AssetWorkspace from "@tonylb/mtw-asset-workspace/dist/"
import { NamespaceMapping } from "@tonylb/mtw-asset-workspace/ts/readOnly"
import { NormalForm } from "@tonylb/mtw-wml/ts/normalize/baseClasses"

type JSONFileCache = {
    normal: NormalForm;
    namespaceIdToDB: NamespaceMapping;
}

export class InheritanceGraph extends Graph<EphemeraAssetId, { key: EphemeraAssetId; address: AssetWorkspaceAddress }, {}> {}

export class FetchImportsJSONHelper {
    _inheritanceGraph: InheritanceGraph
    constructor(inheritanceGraph: InheritanceGraph) {
        this._inheritanceGraph = inheritanceGraph
    }

    async get(assetId: EphemeraAssetId): Promise<JSONFileCache> {
        const node = this._inheritanceGraph.nodes[assetId]
        if (node) {
            const assetWorkspace = new AssetWorkspace(node.address)
            await assetWorkspace.loadJSON()
            return {
                normal: assetWorkspace.normal as NormalForm || {},
                namespaceIdToDB: assetWorkspace.namespaceIdToDB || []
            }    
        }
        else {
            return {
                normal: {},
                namespaceIdToDB: []
            }
        }
    }
}

