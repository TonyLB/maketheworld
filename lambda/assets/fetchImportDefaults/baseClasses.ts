import { EphemeraAssetId } from "@tonylb/mtw-interfaces/ts/baseClasses"
import { JSONFileCache } from "../internalCache/jsonFile"
import { Graph } from "@tonylb/mtw-utilities/dist/graphStorage/utils/graph"
import { AssetWorkspaceAddress } from "@tonylb/mtw-asset-workspace"
import ReadOnlyAssetWorkspace from "@tonylb/mtw-asset-workspace/dist/readOnly"

export class InheritanceGraph extends Graph<EphemeraAssetId, { key: EphemeraAssetId; address: AssetWorkspaceAddress }, {}> {}

export class FetchImportsJSONHelper {
    _inheritanceGraph: InheritanceGraph
    constructor(inheritanceGraph: InheritanceGraph) {
        this._inheritanceGraph = inheritanceGraph
    }

    async get(assetId: EphemeraAssetId): Promise<JSONFileCache> {
        const node = this._inheritanceGraph.nodes[assetId]
        if (node) {
            const assetWorkspace = new ReadOnlyAssetWorkspace(node.address)
            await assetWorkspace.loadJSON()
            return {
                normal: assetWorkspace.normal || {},
                namespaceIdToDB: assetWorkspace.namespaceIdToDB || {}
            }    
        }
        else {
            return {
                normal: {},
                namespaceIdToDB: {}
            }
        }
    }
}

