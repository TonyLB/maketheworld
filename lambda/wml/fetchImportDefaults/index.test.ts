import Normalizer from '@tonylb/mtw-wml/dist/normalize'
import { fetchImports } from '.'

jest.mock('../clients')
import { apiClient } from '../clients'
jest.mock('./baseClasses')
import { FetchImportsJSONHelper } from './baseClasses'
import { NormalForm } from '@tonylb/mtw-wml/dist/normalize/baseClasses'
import { Graph } from '@tonylb/mtw-utilities/dist/graphStorage/utils/graph'
import { EphemeraAssetId } from '@tonylb/mtw-interfaces/ts/baseClasses'
import { AssetWorkspaceAddress } from '@tonylb/mtw-asset-workspace'

const apiClientMock = apiClient as jest.Mocked<typeof apiClient>

const testNormalFromWML = (wml: string): NormalForm => {
    const normalizer = new Normalizer()
    normalizer.loadWML(wml)
    return normalizer.normal
}

describe('fetchImports', () => {
    const testFinal = testNormalFromWML(`<Asset key=(testFinal)>
        <Room key=(testNonImport)>
            <Description>
                DescriptionOne
            </Description>
            <Exit to=(testNonImportStub)>test exit</Exit>
        </Room>
        <Room key=(testNonImportStub)>
            <Name>StubOne</Name>
        </Room>
        <Room key=(testImportOne)>
            <Description>
                Two
            </Description>
            <Exit to=(testImportStubOne)>test exit one</Exit>
        </Room>
        <Room key=(testImportStubOne) />
        <Room key=(testImportTwo) />
        <Room key=(testNonImportTwo)>
            <Exit to=(testImportTwo)>test exit</Exit>
        </Room>
        <Room key=(testImportThree) />
        <Import from=(testImportAssetOne)>
            <Use type="Room" key=(testImportOne) />
            <Use type="Room" key=(testImportStubOne) />
            <Use type="Room" key=(testImportFoo) as=(testImportTwo) />
        </Import>
        <Import from=(testImportAssetTwo)>
            <Use type="Room" key=(basic) as=(testImportThree) />
        </Import>
        <Room key=(testFeatures)>
            <Description>
                <Link to=(featureImport)>Test</Link>
            </Description>
        </Room>
        <Import from=(testImportAssetFour)>
            <Use type="Feature" key=(testFeature) as=(featureImport) />
            <Use type="Room" key=(testRoomWithFeatures) />
        </Import>
    </Asset>`)
    const testImportOne = testNormalFromWML(`<Asset key=(testImportAssetOne)>
        <Room key=(testImportOne)>
            <Description>
                One
            </Description>
        </Room>
        <Room key=(testImportStubOne)>
            <Name>StubTwo</Name>
        </Room>
        <Room key=(testImportFoo)>
            <Name>StubFoo</Name>
            <Description>
                Foo
            </Description>
        </Room>
    </Asset>`)
    const testImportTwo = testNormalFromWML(`<Asset key=(testImportAssetTwo)>
        <Room key=(basic)>
            <Description>
                Asset Two
            </Description>
            <Exit to=(stub)>test exit</Exit>
        </Room>
        <Room key=(stub) />
        <Import from=(testImportAssetThree)>
            <Use type="Room" key=(basicTwo) as=(stub) />
            <Use type="Room" key=(basicOne) as=(basic) />
        </Import>
    </Asset>`)
    const testImportThree = testNormalFromWML(`<Asset key=(testImportAssetThree)>
        <Room key=(basicOne)>
            <Exit to=(stub)>test exit</Exit>
        </Room>
        <Room key=(basicTwo)><Name>Asset Three</Name></Room>
        <Room key=(stub)><Name>AssetThreeStub</Name></Room>
    </Asset>`)
    const testImportFour = testNormalFromWML(`<Asset key=(testImportAssetFour)>
        <Feature key=(testFeature)>
            <Description>Feature test</Description>
        </Feature>
        <Room key=(testRoomWithFeatures)>
            <Description><Link to=(testFeature)>Test</Link></Description>
        </Room>
    </Asset>`)
    const inheritanceGraph = new Graph<EphemeraAssetId, { key: EphemeraAssetId; address: AssetWorkspaceAddress }, {}>(
        {
            'ASSET#testImportAssetOne': { key: 'ASSET#testImportAssetOne', address: { zone: 'Canon', fileName: 'testOne' }},
            'ASSET#testImportAssetTwo': { key: 'ASSET#testImportAssetTwo', address: { zone: 'Canon', fileName: 'testTwo' }},
            'ASSET#testImportAssetThree': { key: 'ASSET#testImportAssetThree', address: { zone: 'Canon', fileName: 'testThree' }},
            'ASSET#testImportAssetFour': { key: 'ASSET#testImportAssetFour', address: { zone: 'Canon', fileName: 'testFour' }},
            'ASSET#testFinal': { key: 'ASSET#testFinal', address: { zone: 'Canon', fileName: 'testFinal' }}
        },
        [
            { from: 'ASSET#testFinal', to: 'ASSET#testImportAssetOne' },
            { from: 'ASSET#testFinal', to: 'ASSET#testImportAssetTwo' },
            { from: 'ASSET#testFinal', to: 'ASSET#testImportAssetFour' },
            { from: 'ASSET#testImportAssetTwo', to: 'ASSET#testImportAssetThree' }
        ],
        { address: {} as any }
    )
    beforeEach(() => {
        jest.clearAllMocks()
        jest.resetAllMocks()
        jest.spyOn(FetchImportsJSONHelper.prototype, 'get').mockImplementation(async (assetId: string) => {
            let normal: NormalForm = {}
            switch(assetId) {
                case 'ASSET#testFinal':
                    normal = testFinal
                    break
                case 'ASSET#testImportAssetOne':
                    normal = testImportOne
                    break
                case 'ASSET#testImportAssetTwo':
                    normal = testImportTwo
                    break
                case 'ASSET#testImportAssetThree':
                    normal = testImportThree
                    break
                case 'ASSET#testImportAssetFour':
                    normal = testImportFour
                    break
            }
            return {
                normal,
                namespaceIdToDB: []
            }
        })
    })

    it('should return empty when passed no keys', async () => {
        await fetchImports({ ConnectionId: '123', RequestId: '456', inheritanceGraph, payloads: [{ assetId: 'ASSET#testFinal', keys: [] }] })
        expect(JSON.parse(apiClientMock.send.mock.calls[0][0].Data)).toMatchSnapshot()
    })

    it('should return element and stubs when passed non-import key', async () => {
        await fetchImports({ ConnectionId: '123', RequestId: '456', inheritanceGraph, payloads: [{ assetId: 'ASSET#testFinal', keys: ['testNonImport'] }] })
        expect(JSON.parse(apiClientMock.send.mock.calls[0][0].Data)).toMatchSnapshot()
    })

    it('should recursive fetch one level of element and stubs when passed import key', async () => {
        await fetchImports({ ConnectionId: '123', RequestId: '456', inheritanceGraph, payloads: [{ assetId: 'ASSET#testFinal', keys: ['testImportOne'] }] })
        expect(JSON.parse(apiClientMock.send.mock.calls[0][0].Data)).toMatchSnapshot()
    })

    it('should follow dynamic renames in imports', async () => {
        await fetchImports({ ConnectionId: '123', RequestId: '456', inheritanceGraph, payloads: [{ assetId: 'ASSET#testFinal', keys: ['testNonImportTwo'] }] })
        expect(JSON.parse(apiClientMock.send.mock.calls[0][0].Data)).toMatchSnapshot()
    })

    it('should import multilevel and avoid colliding stub names', async () => {
        await fetchImports({ ConnectionId: '123', RequestId: '456', inheritanceGraph, payloads: [{ assetId: 'ASSET#testFinal', keys: ['testImportThree'] }] })
        expect(JSON.parse(apiClientMock.send.mock.calls[0][0].Data)).toMatchSnapshot()
    })

    it('should properly stub out features in room description', async () => {
        await fetchImports({ ConnectionId: '123', RequestId: '456', inheritanceGraph, payloads: [{ assetId: 'ASSET#testFinal', keys: ['testRoomWithFeatures'] }] })
        expect(JSON.parse(apiClientMock.send.mock.calls[0][0].Data)).toMatchSnapshot()
    })

})