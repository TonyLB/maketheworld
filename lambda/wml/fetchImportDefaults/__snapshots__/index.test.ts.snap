// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchImports should follow dynamic renames in imports 1`] = `
{
  "importsByAsset": [
    {
      "assetId": "ASSET#testFinal",
      "wml": "<Asset key=(testFinal)>
    <Inherited>
        <Room key=(testImportTwo)><Name>StubFoo</Name></Room>
        <Room key=(testNonImportTwo)>
            <Exit to=(testImportTwo)>test exit</Exit>
        </Room>
    </Inherited>
</Asset>",
    },
  ],
  "messageType": "FetchImports",
}
`;

exports[`fetchImports should import multilevel and avoid colliding stub names 1`] = `
{
  "importsByAsset": [
    {
      "assetId": "ASSET#testFinal",
      "wml": "<Asset key=(testFinal)>
    <Inherited>
        <Room key=(testImportAssetThree.stub)>
            <Name>AssetThreeStub</Name>
        </Room>
        <Room key=(testImportAssetTwo.stub)><Name>Asset Three</Name></Room>
        <Room key=(testImportThree)>
            <Description>Asset Two</Description>
            <Exit to=(testImportAssetThree.stub)>test exit</Exit>
            <Exit to=(testImportAssetTwo.stub)>test exit</Exit>
        </Room>
    </Inherited>
</Asset>",
    },
  ],
  "messageType": "FetchImports",
}
`;

exports[`fetchImports should properly stub out features in room description 1`] = `
{
  "importsByAsset": [
    {
      "assetId": "ASSET#testFinal",
      "wml": "<Asset key=(testFinal)>
    <Inherited>
        <Room key=(testRoomWithFeatures)>
            <Description><Link to=(testFeature)>Test</Link></Description>
        </Room>
        <Feature key=(testFeature) />
    </Inherited>
</Asset>",
    },
  ],
  "messageType": "FetchImports",
}
`;

exports[`fetchImports should recursive fetch one level of element and stubs when passed import key 1`] = `
{
  "importsByAsset": [
    {
      "assetId": "ASSET#testFinal",
      "wml": "<Asset key=(testFinal)>
    <Inherited>
        <Room key=(testImportOne)>
            <Description>OneTwo</Description>
            <Exit to=(testImportStubOne)>test exit one</Exit>
        </Room>
        <Room key=(testImportStubOne)><Name>StubTwo</Name></Room>
    </Inherited>
</Asset>",
    },
  ],
  "messageType": "FetchImports",
}
`;

exports[`fetchImports should return element and stubs when passed non-import key 1`] = `
{
  "importsByAsset": [
    {
      "assetId": "ASSET#testFinal",
      "wml": "<Asset key=(testFinal)>
    <Inherited>
        <Room key=(testNonImport)>
            <Description>DescriptionOne</Description>
            <Exit to=(testNonImportStub)>test exit</Exit>
        </Room>
        <Room key=(testNonImportStub)><Name>StubOne</Name></Room>
    </Inherited>
</Asset>",
    },
  ],
  "messageType": "FetchImports",
}
`;

exports[`fetchImports should return empty when passed no keys 1`] = `
{
  "importsByAsset": [
    {
      "assetId": "ASSET#testFinal",
      "wml": "<Asset key=(testFinal)><Inherited /></Asset>",
    },
  ],
  "messageType": "FetchImports",
}
`;
